Add feature switch attribute design doc

sbomer opened the pull request:
Opening this as a draft for early review. I wanted to get a few eyes on this before making an API proposal.

edit: moved to https://github.com/dotnet/designs/pull/305

sbomer committed:
Add feature switch attribute design doc

gfoidl commented on docs/design/tools/illink/feature-switch-attributes.md:
```diff
@@ -0,0 +1,482 @@
+# Attribute-based model for feature switches
+
+.NET has [feature switches](https://github.com/dotnet/designs/blob/main/accepted/2020/feature-switches.md) which can be set to turn on/off areas of functionality in our libraries, with optional support for removing unused features when trimming or native AOT compiling. What we describe overall as "feature switches" have many pieces which fit together to enable this:
```
```suggestion
.NET has [feature switches](https://github.com/dotnet/designs/blob/main/accepted/2020/feature-switch.md) which can be set to turn on/off areas of functionality in our libraries, with optional support for removing unused features when trimming or native AOT compiling. What we describe overall as "feature switches" have many pieces which fit together to enable this:
```

gfoidl commented on docs/design/tools/illink/feature-switch-attributes.md:
```diff
@@ -0,0 +1,482 @@
+- **static boolean property**
+- **Requries attributes**
+
+The bold pieces are the focus of this document. [Feature switches](https://github.com/dotnet/designs/blob/main/accepted/2020/feature-switches.md) describes how settings flow from the MSBuild property through the AppContext (for runtime feature checks) or `ILLink.Substitutions.xml` (for feature settings baked-in when trimming). This document aims to describe an attribute-based model to replace som ...
```
```suggestion
The bold pieces are the focus of this document. [Feature switches](https://github.com/dotnet/designs/blob/main/accepted/2020/feature-switch.md) describes how settings flow from the MSBuild property through the AppContext (for runtime feature checks) or `ILLink.Substitutions.xml` (for feature settings baked-in when trimming). This document aims to describe an attribute-based model to replace some of the functionality currently implemented via ILLink.Substitutions.xml, used for branch elimination in ILLink and ILCompiler to remove branches that call into `Requires`-annotated code when trimming.
```

MichalStrehovsky commented on docs/design/tools/illink/feature-switch-attributes.md:
```diff
@@ -0,0 +1,482 @@
+class FeatureGuardAttribute<T> : Attribute
+  where T : Attribute, IFeatureAttribute<T> {}
+
+interface IFeatureAttribute<TSelf> where TSelf : Attribute {
+    static abstract string FeatureName { get; }
+}
```
Could this be a custom attribute instead of an interface?

1. For non-execution scenarios it's an annoyance having to resolve virtual calls. The interface method could be implemented on a base, or by a default interface method and the analysis tool needs to deal with all of that just to get a string out. Custom attribute has it in the blob.
2. Static virtuals and generic attributes will not compile on netstandard2.0. Might not be a concern for trimming/AOT, but could be a concern for some other feature.

vitek-karas replied:
+1 on this - as much as I would love to be able to use all the cool new things, the inability to compile this for `netstandard2.0` is pretty bad.
Using an attribute would effectively provide the same information. We would have to enforce its presence via analysis though.

vitek-karas replied:
The point are libraries which multi-target (our recommended approach). If such a library introduces its own feature guard/switch, it would need to use this. Some authors don't like ifdefs, so they may want to backfill the attributes, which would then not work for netstandard2.0. It's probably not very common, but still worth consideration.

The robustness and ease of the tooling is a bigger argument for me, and since this is not going to be common, we don't need the API to be super clean and super easy to use.

I don't feel strongly about this. I do agree with Michal's concerns though.

MichalStrehovsky replied:
The interface approach introduces other class of robustness issues - there's nothing to enforce that the implementation of the method is a string literal, so we'd either need another analyzer for this to ensure string literal, or live with the fact that it might not be a string literal and have a way to report that.

Other thing is that the Roslyn analyzer wouldn't have visibility into the method body if it's in a different compilation unit.

MichalStrehovsky commented on docs/design/tools/illink/feature-switch-attributes.md:
```diff
@@ -0,0 +1,482 @@
+
+- Take into account how this would interact with an attribute-based model for feature switches
+
+  We will explore what an attribtute-based model for feature switches would look like to ensure that it interacts well with a model for feature guards. It's possible that we would design both in conjunction if they are naturally related.
```
I agree that even if we don't actually implement the switch model, it would be great to have the design for it fully fleshed out. These feel very related.

vitek-karas commented on docs/design/tools/illink/feature-switch-attributes.md:
````diff
@@ -0,0 +1,482 @@
+}
+```
+
+We could use this model even for feature switches similar to `StartupHookSupport`, which don't currently have a `StartupHoopSupportAttribute`. There's no need to actually annotate related APIs with the attribute if that is overkill for a small feature. In this case the attribute definition would just serve as metadata that allows us to define a feature switch in a uniform way.
````
On the surface this doesn't sound right to me - declaring an attribute which will do nothing when applied to something. 
On the other hand - it "might" work if we introduce the attribute as internal - in that case nobody would be able to apply it to anything anyway, but it would still serve the purpose as the ID of the feature.
Or similarly make the constructor of the attribute private -> same effect.

I'm still not sure I like this, but it would work.

One thing I do like about this - we're not using strings to identify things, we use IL metadata - which is nice.

vitek-karas commented on docs/design/tools/illink/feature-switch-attributes.md:
```diff
@@ -0,0 +1,482 @@
+
+We could use this model even for feature switches similar to `StartupHookSupport`, which don't currently have a `StartupHoopSupportAttribute`. There's no need to actually annotate related APIs with the attribute if that is overkill for a small feature. In this case the attribute definition would just serve as metadata that allows us to define a feature switch in a uniform way.
+
+This is fundamentally the same idea outlined in https://github.com/dotnet/designs/pull/261. The hope is that this document provides the motivation for using a unified representation for the attributes that are related to feature switches and trim/AOT analysis.
```
Would be good to have a short section comparing the two specifically and possibly discussing a way to unify them (or even if we need that?)
Also some reasoning why we choose not to use the terminology from that PR instead of "feature" (note that "feature switch" is not actually encoded in any public API or MSBuild property - yet, it's just a concept, changing the terminology would not break anything)

vitek-karas replied:
I would also like to see some discussion of what possibilities there are to implement this incrementally. For example, can we introduce feature guards alone - would it be a breaking change to add the rest later on?

sbomer committed:
PR feedback

- Fix typos
- Move feature guard example earlier
- Replace feature guard notes with separate section on constant prop
- Add comparison with capability API draft, including terminology
- Add implementation notes section

sbomer committed:
Avoid interfaces, generic attributes

sbomer committed:
Add more comparison with platform analyzer

sbomer committed:
Add notes about attribute requirement

vitek-karas commented on docs/design/tools/illink/feature-switch-attributes.md:
```diff
@@ -313,39 +311,115 @@ However, any attribute-based model that we pick to represent features should be
-class FeatureGuardAttribute<T> : Attribute
-  where T : Attribute, IFeatureAttribute<T> {}
+[AttributeUsage(AttributeTargets.Class, Inherited = false)]
+abstract class FeatureAttribute : Attribute { }
```
Worth mentioning - this would effectively provide a way to define custom `Requires...` attributes. Something we've discussed several times in the past. We should note that it's not part of this proposal to actually make that work. This will lead to somewhat weird place where if I define my own feature using these attributes, I would have the guard and everything work correctly, but I still would not get any warnings calling the `Requires..` annotated methods outside of the checked scope.

vitek-karas commented on docs/design/tools/illink/feature-switch-attributes.md:
```diff
@@ -0,0 +1,679 @@
+
+### Validating correctness of feature guards
+
+Better would be for the analyzer to validate that the implementation of a feature check includes a check for `IsDynamicCodeSupported`. This could be done using the analyzer infrastructure we have in place without much cost for simple cases:
+
```
There might be a need to do this type of analysis even in ILLink/ILCompiler, at least eventually. Especially if we start relying on `FeatureGuard` attributes in those tools. I don't think it's valid to rely on the analyzer for validating correctness of an app (even if for just a small portion of it). The one advantage here would be that the `FeatureGuard` attribute is introduced along side the corresponding analyzer support, but still.

vitek-karas commented on docs/design/tools/illink/feature-switch-attributes.md:
```diff
@@ -0,0 +1,679 @@
+
+  The analyzer would need to hard-code the fact that `RuntimeFeature.IsDynamicCodeSupported` corresponds to `RequiresDynamicCodeAttribute`, unless we model this relationship via attributes that represent feature switches.
+
+  This would be sufficient for ILLink/ILCompiler to treat `IsSupported` as a constant based on the feature switch name, assuming it has existing knowledge of the fact that `RuntimeFeature.IsDynamicCodeSupported` is contrelled by the feature switch named `"System.Runtime.CompilerServices.RuntimeFeature.IsDynamicCodeSupported"`, either from a substitution XML or from a separate attribute that encod ...
```
```suggestion
  This would be sufficient for ILLink/ILCompiler to treat `IsSupported` as a constant based on the feature switch name, assuming it has existing knowledge of the fact that `RuntimeFeature.IsDynamicCodeSupported` is controlled by the feature switch named `"System.Runtime.CompilerServices.RuntimeFeature.IsDynamicCodeSupported"`, either from a substitution XML or from a separate attribute that encodes this.
```

vitek-karas commented on docs/design/tools/illink/feature-switch-attributes.md:
```diff
@@ -0,0 +1,679 @@
+
+We rely on trim warnings to alert the app author to the problem, and also default `StartupHookSupport` to `false` for trimmed apps. If we have an attribute-based model for feature guards, we may want to consider inferring these defaults from the `FeatureGuard` instead (so a guard for `RequiresUnreferencedCode` that is also a feature switch would be `false` by default whenever "unreferenced code"  ...
+
+The proposal for now is not to infer any defaults and just take care to set appropriate defaults in the SDK. This means that custom feature guards that are also feature switches will need to do the same. For example, a library that defines a feature switch which guards `RequiresUnreferencedCode` will need to ship with MSBuild targets that disable the feature by default for trimmed apps.
```
I think we should add a sample for this - this part is probably the most complex topic in this document. Would be great to have a sample over startup hook, or something similar. How you tie a guard to a different feature switch and how it's supposed to work.

The defaults discussions is actually only important for illink/ilcompiler. The analyzer shouldn't really care as it needs to analyze both branches anyway. Also the correctness of this is not a problem because the "true" branch were the warnings are suppressed would still produce warnings in the app (illink/ilcompiler) if the feature switch is enabled. We basically need to be in a place where analyzer doesn't produce warnings in any case, but trimmer will produce them depending on the feature switch value.

sbomer committed:
PR feedback

- FeatureAttribute -> RequiresFeatureAttribute
- Add notes about analyzer support for Requires attributes
- Mention feature guard validation in ILLink/ILCompiler
- Expand on inferred defaults for feature guards/switches
